### PR TITLE
Support __len__ in xyz files

### PR DESCRIPTION
**Pull request by ctk3b**
Resolves: #992

### CODE REVIEW

**Comment by swails** [comment]
Won't this exhaust the file iterator and prevent actually *parsing* the file?

**Comment by marscher** [comment]
yes, it will. Probably it is better to create a helper method creating the iter and invoke that again in the length method.

**Comment by ctk3b** [comment]
True. So would the best workaround be to open a new filehandle, iterate through that and report that?

**Comment by marscher** [comment]
exactly

**Comment by swails** [comment]
> True. So would the best workaround be to open a new filehandle, iterate through that and report that?

That's one option.  Another is to only support `__len__` on file handles that support `seek` and `tell` (then you can `tell()`, iterate through, then `seek()` back to where you told at the beginning).  This also only works on files that have not started iterating through already (i.e., `__len__` would change based on doing something like `iterload`).

Using a new file handle is probably the easiest way to do this, albeit possibly less efficient than some alternatives (that don't need to actually *parse* the file in Python, but can rather just look for newlines or byte counts).  But you know what they say about premature optimization...

**Comment by marscher** [comment]
You should also cache this value, once you have computed it.

**Comment by swails** [comment]
> You should also cache this value, once you have computed it.

+1. Easy optimization that reduces potential filesystem thrashing

**Comment by marscher** [comment]
:+1:

**Comment by rmcgibbo** [comment]
Can you add a quick test?

File: mdtraj/formats/xyzfile.py [hunk 1]
```diff
@@ -383,4 +384,12 @@ def tell(self):
 
     def __len__(self):
         """Number of frames in the file. """
-        raise NotImplementedError()
+        if str(self._mode) != 'r':
+            raise NotImplementedError('len() only available in mode="r" currently')
+        if not self._is_open:
```
Comment (marscher) [inline mdtraj/formats/xyzfile.py]: If you're opening a new handle, why do require the parsers file handle to be opened?

Comment (rmcgibbo) [inline mdtraj/formats/xyzfile.py]: the fact that it opens a new handle should be considered an implementation detail.

**Comment by rmcgibbo** [comment]
Thanks, guys!

**Pull request merged**